### PR TITLE
KAFKA-14661: Upgrade Zookeeper to 3.8.1

### PR DESCRIPTION
**Pull request by clolov**
Implementation of https://cwiki.apache.org/confluence/display/KAFKA/KIP-902%3A+Upgrade+Zookeeper+to+3.8.1

### CODE REVIEW

**Comment by clolov** [comment]
Hello @mimaison, @ijuma and @cmccabe! Since [KIP-902](https://cwiki.apache.org/confluence/display/KAFKA/KIP-902%3A+Upgrade+Zookeeper+to+3.8.1) has been accepted I have rebased this pull request and added to the documentation for 3.6. I would be grateful for all suggestions for improvements.

**Comment by mimaison** [comment]
Thanks @clolov for the PR. I think some system tests do upgrades from previous versions to the latest. Have you checked these system tests still work with this change?

**Comment by clolov** [comment]
I have not, but I will aim to carry out the check in the upcoming days.

**Comment by clolov** [comment]
Heya @mimaison! I have been struggling with setting up the system test infrastructure as detailed in https://github.com/apache/kafka/tree/trunk/tests - I am trying to run them on EC2. Is there someone you know who has successfully set things up so I can speak with them?

**Comment by mimaison** [comment]
I've not setup the system tests recently. If the provided instructions are not clear enough, let's discuss it on the dev mailing list. You can also open Jiras for the issues you encounter.

**Comment by clolov** [comment]
I ran https://github.com/apache/kafka/blob/trunk/tests/kafkatest/tests/core/upgrade_test.py#L33 on a cluster of 7 m5.xlarge EC2 instances and all tests passed. I have confirmed that as part of the tests we start with an older version of Zookeeper
```
example jinfo output on Zookeeper instances
...
/opt/kafka-3.3.1/bin/../libs/zookeeper-3.6.3.jar
```
and we end up with a new one
```
example jinfo output on Zookeeper instances
...
/opt/kafka-dev/bin/../core/build/dependant-libs-2.13.10/zookeeper-3.8.1.jar
```
I was expecting upgrade tests to fail as detailed in the KIP, but they haven't done this. I will run https://github.com/apache/kafka/blob/trunk/tests/kafkatest/tests/core/downgrade_test.py#L25 as well and report back my findings.

**Comment by clolov** [comment]
Heya @mimaison! I ran the upgrade/downgrade tests over the weekend and they passed as well. Unless there is something else which you would like me to run (or there is a place where you would like me to upload the results from the tests) I believe this is ready to be merged

**Review by divijvaidya — APPROVED** [review]
Thanks Christo! The change looks good to me. Let's wait to hear from @mimaison before we merge this in.

@mimaison please let us know if you have any pending concerns here. I think that we should merge this in soon to ensure adequate bake time before 3.6 release.

**Comment by mimaison** [comment]
I'm not familiar with these tests so please correct me if I'm not understanding something correctly.

First, the KIP states (Compatibility, Deprecation, and Migration Plan section):
```
Users of Kafka clusters with Zookeeper clients older than 3.5.x won't be able to communicate with a 3.8.1 Zookeeper cluster.
```

Now looking at the tests in `upgrade_test.py`, `test_upgrade()` tries deploying a Kafka cluster and ZooKeeper from all previous releases and upgrade them to the latest version. 

So for example, it deploys a 0.10.0.0 Kafka cluster with ZooKeeper 3.4.6. Then proceeds to upgrade ZooKeeper to the version from the latest Kafka which in this case should be 3.8.1. At this stage we should be in the scenario described in the KIP, Kafka brokers (still running 0.10.0.0) have ZooKeeper clients older than 3.5.x and have to communicate with ZooKeeper 3.8.1 which should not work.

Can you provide the output you got when running these tests?

**Comment by divijvaidya** [comment]
FYI @mimaison , @clolov is on vacation this week. I will bring this comment to his attention when he is back next week.

**Comment by mimaison** [comment]
Also we need to update the `LICENSE-binary` file as mentioned in https://github.com/apache/kafka/blob/trunk/gradle/dependencies.gradle#L73-L74

**Comment by clolov** [comment]
> I'm not familiar with these tests so please correct me if I'm not understanding something correctly.
> 
> First, the KIP states (Compatibility, Deprecation, and Migration Plan section):
> 
> ```
> Users of Kafka clusters with Zookeeper clients older than 3.5.x won't be able to communicate with a 3.8.1 Zookeeper cluster.
> ```
> 
> Now looking at the tests in `upgrade_test.py`, `test_upgrade()` tries deploying a Kafka cluster and ZooKeeper from all previous releases and upgrade them to the latest version.
> 
> So for example, it deploys a 0.10.0.0 Kafka cluster with ZooKeeper 3.4.6. Then proceeds to upgrade ZooKeeper to the version from the latest Kafka which in this case should be 3.8.1. At this stage we should be in the scenario described in the KIP, Kafka brokers (still running 0.10.0.0) have ZooKeeper clients older than 3.5.x and have to communicate with ZooKeeper 3.8.1 which should not work.
> 
> Can you provide the output you got when running these tests?

Heya @mimaison! Yes, as mentioned in a previous comment of mine, I have confirmed that we start with an older version of Zookeeper and as part of the tests we upgrade to the latest. The KIP was written based on information from Zookeeper's official documentation as I was unaware of these tests at the time. I suspect the tests do not fail because the Zookeeper APIs Kafka uses are compatible across all versions despite what Zookeeper's documentation says. This being said, these upgrade/downgrade tests only test production + consumption so I would rather be safe than sorry and go forward with recommending to do a double-rolling-upgrade as documented in the KIP. The current output of the tests is
```
ubuntu@ip-A-X-Y-Z:~/kafka/results$ du -sh -- */ | sort -rh
50G	2023-06-23--003/
50G	2023-06-22--002/
```
so I suspect it will be difficult to share via GitHub. Where would you like me to put the results?

**Comment by forlack** [comment]
Should we bump to ZK 3.8.2 to include https://issues.apache.org/jira/browse/ZOOKEEPER-3860?  Have seen a few kafka customers hit this issue recently.

**Comment by clolov** [comment]
Hey @forlack! Thank you for the suggestion, I have upgraded the dependency so that tests can run.

Heya @mimaison, I have addressed your comment about adding the latest version to the LICENSE-binary file and I would be grateful for any last review as this has been open for sometime.

**Comment by mimaison** [comment]
My point is that this really contradicts the KIP. In the KIP we say "it does not work" but then we have a test that relies on the upgrade working. This is a bit counter intuitive.

**Comment by clolov** [comment]
Heya @mimaison! Okay, I can think of and propose two ways forward:

1. I go back and modify the KIP based on the outcome of these tests to say it is safe to carry out the upgrade without a "bridge" upgrade.

2. We treat the tests passing as a happy-path upgrade and continue recommending a "bridge" upgrade.

As far as I know the upgrade/downgrade tests do not test all interactions that Kafka has with Zookeeper, so asking people to carry out two upgrades won't break their clusters even if only one is needed. As mentioned, the statement made in the KIP is a statement based on Zookeeper's documentation.

As such, I have a preference for 2. What would you like to go for? If you disagree with both what would your recommendation for going forward be?

**Review by mimaison — APPROVED** [review]
If the ZooKeeper documentation says there are some incompatibilities, let's not go with option 1. The upgrade test still seems a bit strange but let's get the new ZooKeeper version in.

**Review by divijvaidya — APPROVED** [review]
Thank you Christo. Waiting for CI to produce sane results. Will merge after that.

**Comment by divijvaidya** [comment]
@clolov we have CI failures. Can you please check what's going on here.

**Comment by clolov** [comment]
It appears that Jenkins is failing on `**/build/test-results/**/TEST-*.xml— Archive JUnit-formatted test results`. There are some failures in the tests, but the step itself has an output of `BUILD SUCCESSFUL in 54m 57s`. I will check whether other recent pull requests appear to have the same failure.

**Comment by clolov** [comment]
At least the below tests have started to continuously fail. I am still looking as to why:
```
Gradle Test Run :core:test > Gradle Test Executor 3 > ControllerIntegrationTest > testControllerMoveOnPreferredReplicaElection() FAILED
    org.opentest4j.AssertionFailedError: expected: <true> but was: <false>
        at app//org.junit.jupiter.api.AssertionFailureBuilder.build(AssertionFailureBuilder.java:151)
        at app//org.junit.jupiter.api.AssertionFailureBuilder.buildAndThrow(AssertionFailureBuilder.java:132)
        at app//org.junit.jupiter.api.AssertTrue.failNotTrue(AssertTrue.java:63)
        at app//org.junit.jupiter.api.AssertTrue.assertTrue(AssertTrue.java:36)
        at app//org.junit.jupiter.api.AssertTrue.assertTrue(AssertTrue.java:31)
        at app//org.junit.jupiter.api.Assertions.assertTrue(Assertions.java:180)
        at app//kafka.controller.ControllerIntegrationTest.testControllerMoveOnPreferredReplicaElection(ControllerIntegrationTest.scala:1885)

Gradle Test Run :core:test > Gradle Test Executor 3 > ControllerIntegrationTest > testControllerMoveOnTopicCreation() FAILED
    org.opentest4j.AssertionFailedError: expected: <true> but was: <false>
        at app//org.junit.jupiter.api.AssertionFailureBuilder.build(AssertionFailureBuilder.java:151)
        at app//org.junit.jupiter.api.AssertionFailureBuilder.buildAndThrow(AssertionFailureBuilder.java:132)
        at app//org.junit.jupiter.api.AssertTrue.failNotTrue(AssertTrue.java:63)
        at app//org.junit.jupiter.api.AssertTrue.assertTrue(AssertTrue.java:36)
        at app//org.junit.jupiter.api.AssertTrue.assertTrue(AssertTrue.java:31)
        at app//org.junit.jupiter.api.Assertions.assertTrue(Assertions.java:180)
        at app//kafka.controller.ControllerIntegrationTest.testControllerMoveOnTopicCreation(ControllerIntegrationTest.scala:1885)

Gradle Test Run :core:test > Gradle Test Executor 3 > ControllerIntegrationTest > testControllerMoveOnPartitionReassignment() FAILED
    org.opentest4j.AssertionFailedError: expected: <true> but was: <false>
        at app//org.junit.jupiter.api.AssertionFailureBuilder.build(AssertionFailureBuilder.java:151)
        at app//org.junit.jupiter.api.AssertionFailureBuilder.buildAndThrow(AssertionFailureBuilder.java:132)
        at app//org.junit.jupiter.api.AssertTrue.failNotTrue(AssertTrue.java:63)
        at app//org.junit.jupiter.api.AssertTrue.assertTrue(AssertTrue.java:36)
        at app//org.junit.jupiter.api.AssertTrue.assertTrue(AssertTrue.java:31)
        at app//org.junit.jupiter.api.Assertions.assertTrue(Assertions.java:180)
        at app//kafka.controller.ControllerIntegrationTest.testControllerMoveOnPartitionReassignment(ControllerIntegrationTest.scala:1885)

Gradle Test Run :core:test > Gradle Test Executor 3 > ControllerIntegrationTest > testControllerMoveOnTopicDeletion() FAILED
    org.opentest4j.AssertionFailedError: expected: <true> but was: <false>
        at app//org.junit.jupiter.api.AssertionFailureBuilder.build(AssertionFailureBuilder.java:151)
        at app//org.junit.jupiter.api.AssertionFailureBuilder.buildAndThrow(AssertionFailureBuilder.java:132)
        at app//org.junit.jupiter.api.AssertTrue.failNotTrue(AssertTrue.java:63)
        at app//org.junit.jupiter.api.AssertTrue.assertTrue(AssertTrue.java:36)
        at app//org.junit.jupiter.api.AssertTrue.assertTrue(AssertTrue.java:31)
        at app//org.junit.jupiter.api.Assertions.assertTrue(Assertions.java:180)
        at app//kafka.controller.ControllerIntegrationTest.testControllerMoveOnTopicDeletion(ControllerIntegrationTest.scala:1885)
```

**Comment by clolov** [comment]
At least on a first look it appears something changed/is wrong with the loggers:
```
[2023-02-22T17:35:34.655Z] Gradle Test Run :core:unitTest > Gradle Test Executor 136 > ConfigAdminManagerTest > testPreprocessIncrementalWithLoggerChanges() FAILED
[2023-02-22T17:35:34.655Z]     org.opentest4j.AssertionFailedError: expected: <{AlterConfigsResource(resourceType=8, resourceName='1', configs=[AlterableConfig(name='kafka.server.ConfigAdminManagerTest', configOperation=0, value='INFO')])=ApiError(error=NONE, message=null)}> but was: <{AlterConfigsResource(resourceType=8, resourceName='1', configs=[AlterableConfig(name='kafka.server.ConfigAdminManagerTest', configOperation=0, value='INFO')])=ApiError(error=INVALID_CONFIG, message=Logger kafka.server.ConfigAdminManagerTest does not exist!)}>
[2023-02-22T17:35:34.655Z]         at app//org.junit.jupiter.api.AssertionFailureBuilder.build(AssertionFailureBuilder.java:151)
[2023-02-22T17:35:34.655Z]         at app//org.junit.jupiter.api.AssertionFailureBuilder.buildAndThrow(AssertionFailureBuilder.java:132)
[2023-02-22T17:35:34.655Z]         at app//org.junit.jupiter.api.AssertEquals.failNotEqual(AssertEquals.java:197)
[2023-02-22T17:35:34.655Z]         at app//org.junit.jupiter.api.AssertEquals.assertEquals(AssertEquals.java:182)
[2023-02-22T17:35:34.655Z]         at app//org.junit.jupiter.api.AssertEquals.assertEquals(AssertEquals.java:177)
[2023-02-22T17:35:34.655Z]         at app//org.junit.jupiter.api.Assertions.assertEquals(Assertions.java:1142)
[2023-02-22T17:35:34.656Z]         at app//kafka.server.ConfigAdminManagerTest.testPreprocessIncrementalWithLoggerChanges(ConfigAdminManagerTest.scala:352)

...

[2023-02-22T17:35:34.656Z] Gradle Test Run :core:unitTest > Gradle Test Executor 136 > ConfigAdminManagerTest > testValidateLogLevelConfigs() FAILED
[2023-02-22T17:35:34.656Z]     org.apache.kafka.common.errors.InvalidConfigurationException: Logger kafka.server.ConfigAdminManagerTest does not exist!

...

[2023-02-22T17:36:10.045Z] Gradle Test Run :core:unitTest > Gradle Test Executor 136 > LoggingTest > testLoggerLevelIsResolved() FAILED
[2023-02-22T17:36:10.045Z]     org.opentest4j.AssertionFailedError: expected: <TRACE> but was: <No such logger.>
[2023-02-22T17:36:10.045Z]         at app//org.junit.jupiter.api.AssertionFailureBuilder.build(AssertionFailureBuilder.java:151)
[2023-02-22T17:36:10.045Z]         at app//org.junit.jupiter.api.AssertionFailureBuilder.buildAndThrow(AssertionFailureBuilder.java:132)
[2023-02-22T17:36:10.045Z]         at app//org.junit.jupiter.api.AssertEquals.failNotEqual(AssertEquals.java:197)
[2023-02-22T17:36:10.045Z]         at app//org.junit.jupiter.api.AssertEquals.assertEquals(AssertEquals.java:182)
[2023-02-22T17:36:10.045Z]         at app//org.junit.jupiter.api.AssertEquals.assertEquals(AssertEquals.java:177)
[2023-02-22T17:36:10.045Z]         at app//org.junit.jupiter.api.Assertions.assertEquals(Assertions.java:1142)
[2023-02-22T17:36:10.045Z]         at app//kafka.utils.LoggingTest.testLoggerLevelIsResolved(LoggingTest.scala:81)

...

[2023-02-22T17:37:30.427Z] Gradle Test Run :core:unitTest > Gradle Test Executor 173 > ConfigAdminManagerTest > testPreprocessIncrementalWithLoggerChanges() FAILED
[2023-02-22T17:37:30.427Z]     org.opentest4j.AssertionFailedError: expected: <{AlterConfigsResource(resourceType=8, resourceName='1', configs=[AlterableConfig(name='kafka.server.ConfigAdminManagerTest', configOperation=0, value='INFO')])=ApiError(error=NONE, message=null)}> but was: <{AlterConfigsResource(resourceType=8, resourceName='1', configs=[AlterableConfig(name='kafka.server.ConfigAdminManagerTest', configOperation=0, value='INFO')])=ApiError(error=INVALID_CONFIG, message=Logger kafka.server.ConfigAdminManagerTest does not exist!)}>
[2023-02-22T17:37:30.427Z]         at app//org.junit.jupiter.api.AssertionFailureBuilder.build(AssertionFailureBuilder.java:151)
[2023-02-22T17:37:30.427Z]         at app//org.junit.jupiter.api.AssertionFailureBuilder.buildAndThrow(AssertionFailureBuilder.java:132)
[2023-02-22T17:37:30.427Z]         at app//org.junit.jupiter.api.AssertEquals.failNotEqual(AssertEquals.java:197)
[2023-02-22T17:37:30.427Z]         at app//org.junit.jupiter.api.AssertEquals.assertEquals(AssertEquals.java:182)
[2023-02-22T17:37:30.427Z]         at app//org.junit.jupiter.api.AssertEquals.assertEquals(AssertEquals.java:177)
[2023-02-22T17:37:30.427Z]         at app//org.junit.jupiter.api.Assertions.assertEquals(Assertions.java:1142)
[2023-02-22T17:37:30.427Z]         at app//kafka.server.ConfigAdminManagerTest.testPreprocessIncrementalWithLoggerChanges(ConfigAdminManagerTest.scala:352)

...

[2023-02-22T17:37:30.427Z] Gradle Test Run :core:unitTest > Gradle Test Executor 173 > ConfigAdminManagerTest > testValidateLogLevelConfigs() FAILED
[2023-02-22T17:37:30.427Z]     org.apache.kafka.common.errors.InvalidConfigurationException: Logger kafka.server.ConfigAdminManagerTest does not exist!

...

[2023-02-22T17:37:30.427Z] Gradle Test Run :core:unitTest > Gradle Test Executor 173 > LoggingTest > testLoggerLevelIsResolved() FAILED
[2023-02-22T17:37:30.427Z]     org.opentest4j.AssertionFailedError: expected: <TRACE> but was: <No such logger.>
[2023-02-22T17:37:30.427Z]         at app//org.junit.jupiter.api.AssertionFailureBuilder.build(AssertionFailureBuilder.java:151)
[2023-02-22T17:37:30.427Z]         at app//org.junit.jupiter.api.AssertionFailureBuilder.buildAndThrow(AssertionFailureBuilder.java:132)
[2023-02-22T17:37:30.427Z]         at app//org.junit.jupiter.api.AssertEquals.failNotEqual(AssertEquals.java:197)
[2023-02-22T17:37:30.427Z]         at app//org.junit.jupiter.api.AssertEquals.assertEquals(AssertEquals.java:182)
[2023-02-22T17:37:30.427Z]         at app//org.junit.jupiter.api.AssertEquals.assertEquals(AssertEquals.java:177)
[2023-02-22T17:37:30.427Z]         at app//org.junit.jupiter.api.Assertions.assertEquals(Assertions.java:1142)
[2023-02-22T17:37:30.427Z]         at app//kafka.utils.LoggingTest.testLoggerLevelIsResolved(LoggingTest.scala:81)

...

[2023-02-22T17:37:30.427Z] 2757 tests completed, 6 failed, 5 skipped
```

**Comment by clolov** [comment]
```
(x2)

[2023-02-22T18:18:23.694Z] Gradle Test Run :core:integrationTest > Gradle Test Executor 175 > PlaintextAdminIntegrationTest > testDescribeConfigsForLog4jLogLevels(String) > kafka.api.PlaintextAdminIntegrationTest.testDescribeConfigsForLog4jLogLevels(String)[1] FAILED
[2023-02-22T18:18:23.694Z]     java.lang.NullPointerException
[2023-02-22T18:18:23.694Z]         at kafka.api.PlaintextAdminIntegrationTest.testDescribeConfigsForLog4jLogLevels(PlaintextAdminIntegrationTest.scala:2260)

...

(x2)

[2023-02-22T18:42:58.206Z] Gradle Test Run :core:integrationTest > Gradle Test Executor 176 > ControllerIntegrationTest > testControllerMoveOnTopicDeletion() FAILED
[2023-02-22T18:42:58.206Z]     org.opentest4j.AssertionFailedError: expected: <true> but was: <false>
[2023-02-22T18:42:58.206Z]         at app//org.junit.jupiter.api.AssertionFailureBuilder.build(AssertionFailureBuilder.java:151)
[2023-02-22T18:42:58.206Z]         at app//org.junit.jupiter.api.AssertionFailureBuilder.buildAndThrow(AssertionFailureBuilder.java:132)
[2023-02-22T18:42:58.206Z]         at app//org.junit.jupiter.api.AssertTrue.failNotTrue(AssertTrue.java:63)
[2023-02-22T18:42:58.206Z]         at app//org.junit.jupiter.api.AssertTrue.assertTrue(AssertTrue.java:36)
[2023-02-22T18:42:58.206Z]         at app//org.junit.jupiter.api.AssertTrue.assertTrue(AssertTrue.java:31)
[2023-02-22T18:42:58.206Z]         at app//org.junit.jupiter.api.Assertions.assertTrue(Assertions.java:180)
[2023-02-22T18:42:58.206Z]         at app//kafka.controller.ControllerIntegrationTest.testControllerMoveOnTopicDeletion(ControllerIntegrationTest.scala:1884)

...

(x2)

[2023-02-22T18:42:58.207Z] Gradle Test Run :core:integrationTest > Gradle Test Executor 176 > ControllerIntegrationTest > testControllerMoveOnPartitionReassignment() FAILED
[2023-02-22T18:42:58.207Z]     org.opentest4j.AssertionFailedError: expected: <true> but was: <false>
[2023-02-22T18:42:58.207Z]         at app//org.junit.jupiter.api.AssertionFailureBuilder.build(AssertionFailureBuilder.java:151)
[2023-02-22T18:42:58.207Z]         at app//org.junit.jupiter.api.AssertionFailureBuilder.buildAndThrow(AssertionFailureBuilder.java:132)
[2023-02-22T18:42:58.207Z]         at app//org.junit.jupiter.api.AssertTrue.failNotTrue(AssertTrue.java:63)
[2023-02-22T18:42:58.207Z]         at app//org.junit.jupiter.api.AssertTrue.assertTrue(AssertTrue.java:36)
[2023-02-22T18:42:58.207Z]         at app//org.junit.jupiter.api.AssertTrue.assertTrue(AssertTrue.java:31)
[2023-02-22T18:42:58.207Z]         at app//org.junit.jupiter.api.Assertions.assertTrue(Assertions.java:180)
[2023-02-22T18:42:58.207Z]         at app//kafka.controller.ControllerIntegrationTest.testControllerMoveOnPartitionReassignment(ControllerIntegrationTest.scala:1884)

...

(x2)

[2023-02-22T18:43:01.009Z] Gradle Test Run :core:integrationTest > Gradle Test Executor 176 > ControllerIntegrationTest > testControllerMoveOnTopicCreation() FAILED
[2023-02-22T18:43:01.009Z]     org.opentest4j.AssertionFailedError: expected: <true> but was: <false>
[2023-02-22T18:43:01.009Z]         at app//org.junit.jupiter.api.AssertionFailureBuilder.build(AssertionFailureBuilder.java:151)
[2023-02-22T18:43:01.009Z]         at app//org.junit.jupiter.api.AssertionFailureBuilder.buildAndThrow(AssertionFailureBuilder.java:132)
[2023-02-22T18:43:01.009Z]         at app//org.junit.jupiter.api.AssertTrue.failNotTrue(AssertTrue.java:63)
[2023-02-22T18:43:01.009Z]         at app//org.junit.jupiter.api.AssertTrue.assertTrue(AssertTrue.java:36)
[2023-02-22T18:43:01.009Z]         at app//org.junit.jupiter.api.AssertTrue.assertTrue(AssertTrue.java:31)
[2023-02-22T18:43:01.009Z]         at app//org.junit.jupiter.api.Assertions.assertTrue(Assertions.java:180)
[2023-02-22T18:43:01.009Z]         at app//kafka.controller.ControllerIntegrationTest.testControllerMoveOnTopicCreation(ControllerIntegrationTest.scala:1884)

...

(x2)

[2023-02-22T18:43:52.201Z] Gradle Test Run :core:integrationTest > Gradle Test Executor 176 > ControllerIntegrationTest > testControllerMoveOnPreferredReplicaElection() FAILED
[2023-02-22T18:43:52.201Z]     org.opentest4j.AssertionFailedError: expected: <true> but was: <false>
[2023-02-22T18:43:52.201Z]         at app//org.junit.jupiter.api.AssertionFailureBuilder.build(AssertionFailureBuilder.java:151)
[2023-02-22T18:43:52.201Z]         at app//org.junit.jupiter.api.AssertionFailureBuilder.buildAndThrow(AssertionFailureBuilder.java:132)
[2023-02-22T18:43:52.201Z]         at app//org.junit.jupiter.api.AssertTrue.failNotTrue(AssertTrue.java:63)
[2023-02-22T18:43:52.201Z]         at app//org.junit.jupiter.api.AssertTrue.assertTrue(AssertTrue.java:36)
[2023-02-22T18:43:52.201Z]         at app//org.junit.jupiter.api.AssertTrue.assertTrue(AssertTrue.java:31)
[2023-02-22T18:43:52.201Z]         at app//org.junit.jupiter.api.Assertions.assertTrue(Assertions.java:180)
[2023-02-22T18:43:52.201Z]         at app//kafka.controller.ControllerIntegrationTest.testControllerMoveOnPreferredReplicaElection(ControllerIntegrationTest.scala:1884)

...

(x2)

[2023-02-22T19:02:05.196Z] Gradle Test Run :core:integrationTest > Gradle Test Executor 176 > KRaftClusterTest > testSetLog4jConfigurations() FAILED
[2023-02-22T19:02:05.196Z]     org.opentest4j.AssertionFailedError: expected: <List(ApiError(error=NONE, message=null), ApiError(error=INVALID_REQUEST, message=APPEND operation is not allowed for the BROKER_LOGGER resource))> but was: <List(ApiError(error=INVALID_CONFIG, message=Logger kafka.server.KRaftClusterTest does not exist!), ApiError(error=INVALID_REQUEST, message=APPEND operation is not allowed for the BROKER_LOGGER resource))>
[2023-02-22T19:02:05.196Z]         at app//org.junit.jupiter.api.AssertionFailureBuilder.build(AssertionFailureBuilder.java:151)
[2023-02-22T19:02:05.196Z]         at app//org.junit.jupiter.api.AssertionFailureBuilder.buildAndThrow(AssertionFailureBuilder.java:132)
[2023-02-22T19:02:05.196Z]         at app//org.junit.jupiter.api.AssertEquals.failNotEqual(AssertEquals.java:197)
[2023-02-22T19:02:05.196Z]         at app//org.junit.jupiter.api.AssertEquals.assertEquals(AssertEquals.java:182)
[2023-02-22T19:02:05.196Z]         at app//org.junit.jupiter.api.AssertEquals.assertEquals(AssertEquals.java:177)
[2023-02-22T19:02:05.196Z]         at app//org.junit.jupiter.api.Assertions.assertEquals(Assertions.java:1142)
[2023-02-22T19:02:05.196Z]         at app//kafka.server.KRaftClusterTest.testSetLog4jConfigurations(KRaftClusterTest.scala:645)

...

(x2)

[2023-02-22T19:33:57.499Z] Gradle Test Run :core:integrationTest > Gradle Test Executor 180 > PlaintextAdminIntegrationTest > testDescribeConfigsForLog4jLogLevels(String) > kafka.api.PlaintextAdminIntegrationTest.testDescribeConfigsForLog4jLogLevels(String)[1] FAILED
[2023-02-22T19:33:57.499Z]     java.lang.NullPointerException
[2023-02-22T19:33:57.499Z]         at kafka.api.PlaintextAdminIntegrationTest.testDescribeConfigsForLog4jLogLevels(PlaintextAdminIntegrationTest.scala:2260)

...

[2023-02-22T19:34:12.144Z] 1970 tests completed, 14 failed, 5 skipped
```

**Comment by clolov** [comment]
Zoning in on ControllerIntegrationTest#testControllerMoveOnTopicCreation:

```
...
1882:      val event = appender.getMessages.find(e => e.getLevel == Level.INFO
1882:        && e.getThrowableInformation != null
1883:        && e.getThrowableInformation.getThrowable.getClass.getName.equals(classOf[ControllerMovedException].getName))
1884:      assertTrue(event.isDefined) <- THIS IS THE ASSERTION WHICH FAILS
...
```

However from the test logs:
```
16:03:02.258 [controller-event-thread] INFO kafka.controller.KafkaController - [Controller id=0] Controller moved to another broker when processing TopicChange.
org.apache.kafka.common.errors.ControllerMovedException: Controller epoch zkVersion check fails. Expected zkVersion = 1
```

**Comment by clolov** [comment]
It appears that in Zookeeper 3.8.0 the Zookeeper community moved away from log4j to logback (source: https://issues.apache.org/jira/browse/ZOOKEEPER-4427)

In the failing test cases we attach an appender to the root logger (Kafka uses a ~log4j~ reload4j logger) and assert that we see certain log messages. Because of the mismatch in expectation (~log4j~ reload4j) and reality (logback) these tests have started failing as the test-appender has no messages.

The difference in Zookeeper dependencies between current trunk (31 July 2023) and this pull request is:
```
trunk

org.apache.zookeeper:zookeeper:3.6.4
> io.netty:netty-handler:4.1.94.Final
> io.netty:netty-transport-native-epoll:4.1.94.Final
> org.apache.yetus:audience-annotations:0.13.0
> org.apache.zookeeper:zookeeper-jute:3.6.4
> org.slf4j:slf4j-api:1.7.36

pull request

org.apache.zookeeper:zookeeper:3.8.2
> ch.qos.logback:logback-classic:1.2.10 <- NEW
> ch.qos.logback:logback-core:1.2.10    <- NEW
> commons-io:commons-io:2.11.0          <- NEW
> io.netty:netty-handler:4.1.94.Final
> io.netty:netty-transport-native-epoll:4.1.94.Final
> org.apache.yetus:audience-annotations:0.12.0
> org.apache.zookeeper:zookeeper-jute:3.8.2
> org.slf4j:slf4j-api:1.7.36
```

I have removed the `ch.qos.logback:*` dependencies in the latest commit and tests have started passing locally.

**Comment by clolov** [comment]
The failures in JDK 8/Scala 2.12 have to do with formatting on code which hasn't been touched as part of this pull request, so I believe it is safe to go ahead.

There are 2 failures in JDK 20/Scala 2.13 - one concerned with tiered storage (RemoteIndexCacheTest - cleaner did not shutdown) and one with simple production (PlaintextProducerSendTest - not enough brokers for specified RF). I believe these are transient failures as they do not manifest when running locally, but if people are unconvinced we can rerun the tests.

Otherwise, I think we are ready to merge this!

File: build.gradle
Comment (divijvaidya) [inline build.gradle:901]: Is this still true with the latest release? (I guess not since commons-cli is explicitly pulled as a dependency of Zk)

File: build.gradle
Comment (divijvaidya) [inline build.gradle]: > but Kafka relies on log4j

this is not correct. Kafka uses slf4j APIs with reload4j as the logging framework. I would suggest to rephrase this as "Zookeeper uses logback as the logging framework but Kafka uses reload4j, hence we want to avoid having multiple frameworks in dependency chain, so removing logback".

Next, can you check why Zk logs are working even if we are using reload4j at runtime? (the answer might be that Zk uses slf4j and underlying framework is pluggable)

Comment (clolov) [reply build.gradle]: Yup, this is the exact reason. Kafka uses slf4j + reload4j. Zookeeper uses slf4j + logback. When we remove logback we fallback to reload4j for both Zookeeper and Kafka. I will update the comment.

File: build.gradle
Comment (clolov) [reply build.gradle:901]: I think commons-cli and commons-io are two different libraries so I would prefer to keep commons-cli.

File: build.gradle
Comment (divijvaidya) [reply build.gradle]: Thank you. That clarifies things. This is a safe change now.

File: build.gradle
Comment (divijvaidya) [reply build.gradle:901]: My bad! you are right.

**Comment by divijvaidya** [comment]
Unrelated test failures at https://ci-builds.apache.org/blue/organizations/jenkins/Kafka%2Fkafka-pr/detail/PR-13260/9/pipeline/ . Merging this one.

**Pull request merged**
